File: backend/schemas/metadata.py
from typing import Literal

from pydantic import BaseModel, Field


class MetadataJobStatus(BaseModel):
    status: Literal["pending", "processing", "completed", "failed"]
    processed_count: int = Field(ge=0)
    total_count: int = Field(ge=0)
    error_message: str | None = None


class MetadataStatusResponse(BaseModel):
    books_with_genres: int = Field(ge=0)
    total_books: int = Field(ge=0)
    job: MetadataJobStatus

File: frontend/src/lib/metadata.ts
import { fetchJson } from "@/lib/api";

export type MetadataJobState = "pending" | "processing" | "completed" | "failed";

export type MetadataStatus = {
  books_with_genres: number;
  total_books: number;
  job: {
    status: MetadataJobState;
    processed_count: number;
    total_count: number;
    error_message?: string | null;
  };
};

export function fetchMetadataStatus(): Promise<MetadataStatus> {
  return fetchJson<MetadataStatus>("/metadata/status", { skipClientCache: true });
}

export function startMetadataGeneration(): Promise<MetadataStatus> {
  return fetchJson<MetadataStatus>("/metadata/generate", {
    method: "POST",
    skipClientCache: true
  });
}

export function metadataStatusLabel(status: MetadataJobState): string {
  if (status === "pending") return "Pending";
  if (status === "processing") return "Processing";
  if (status === "failed") return "Failed";
  return "Completed";
}
